src/serac/physics/utilities/finite_element_vector.hpp

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file finite_element_vector.hpp
+ *
+ * @brief This file contains the declaration of structure that manages vectors
+ * derived from an MFEM finite element space
+ */
+
+#pragma once
+
+#include "mfem.hpp"
+
+#include "serac/physics/utilities/variant.hpp"
+
+namespace serac {
+
+/**
+ * @brief Class for encapsulating the data associated with a vector derived
+ * from a MFEM finite element space. Specifically, it contains the information
+ * needed for both primal finite element state fields and dual finite element vectors.
+ *
+ * Namely: Mesh, FiniteElementCollection, FiniteElementVector,
+ * GridFunction, and a distributed vector of the solution
+ */
+class FiniteElementVector {
+public:
+  /**
+   * @brief Structure for optionally configuring a FiniteElementVector
+   */
+  // The optionals are explicitly default-constructed to allow the user to partially aggregrate-initialized
+  // with only the options they care about
+  struct Options {
+    /**
+     * @brief The polynomial order that should be used for the problem
+     */
+    int order = 1;
+    /**
+     * @brief The number of copies of the finite element collections (e.g. vector_dim = 2 or 3 for solid mechanics).
+     * Defaults to scalar valued spaces.
+     */
+    int vector_dim = 1;
+    /**
+     * @brief The FECollection to use - defaults to an H1_FECollection
+     */
+    std::unique_ptr<mfem::FiniteElementCollection> coll = {};
+    /**
+     * The DOF ordering that should be used interally by MFEM
+     */
+    mfem::Ordering::Type ordering = mfem::Ordering::byVDIM;
+    /**
+     * @brief The name of the field encapsulated by the state object
+     */
+    std::string name = "";
+    /**
+     * @brief Whether the GridFunction should be allocated (and owned by the FEState object)
+     * @note This should only be false for restart-based runs
+     */
+    bool alloc_local = true;
```

Comment:
if the value of this variable is entirely dependent on whether or not it's a restart, it might be clearer to make it something like `bool is_a_restart` (would have to be negated from its current interpretation) or an enum if there are more options

```
// this is a terrible name
enum class SimulationType { Normal, Restart, SomethingElse };
```

Reply:
it would be good to get other people's input on this name though (since I won't pretend to know what it's used for)

Reply:
Yeah, this name is confusing. After digging a bit, I realized "only restart" was also misleading. This is really denoting whether the memory is managed by sidre or not. If it isn't, we prefer to use unique pointers. If it is, they expect a raw new. Would the name `managed_by_sidre` with appropriate comments help?

Reply:
Fixed! Hopefully this makes more sense now.